### PR TITLE
Propagate labels from Pipeline/Task to PipelineRun/TaskRun

### LINKED ISSUE #501: Propagate labels from Pipeline and Task to PipelineRun and TaskRun
## Expected Behavior

Custom labels specified on `Pipeline` and `Task` resources should be added to `PipelineRun` and `TaskRun` objects that reference them. In other words, labels should propagate as follows:

Pipeline -> PipelineRun -> TaskRun -> Pod
Task -> TaskRun -> Pod

In addition, similarly to how the `pipeline.knative.dev/pipeline` label is added to generated `TaskRuns`, a `pipeline.knative.dev/task` label should be added to generated `Pods`.

## Actual Behavior

Custom labels on `Pipeline` and `Task` resources are not propagated to other resources. Labels propagate as follows:

PipelineRun -> TaskRun -> Pod
TaskRun -> Pod

## Additional Info

See #458. #488 added label propagation from `PipelineRun` to `TaskRun` to `Pod`, but did not handle propagation for `Pipeline` and `Task`. Propagating `Pipeline` and `Task` labels will require updating a few data structures so that either the `ObjectMeta` field or the labels field specifically of `Pipelines` and `Tasks` is available when `TaskRuns` and `Pods` are created so that they can be propagated correctly.

Consistent label propagation is desirable because it allows users to easily find all resources associated with a given Pipeline/Task.

Open questions:

* Should labels be propagated from/to resources as well?

### PR DESCRIPTION
**Pull request by dwnusbaum**
## Changes

With this change, labels are propagated from Pipeline and Task to
PipelineRun and TaskRun, respectively, giving us full label propagation
from Pipeline to PipelineRun to TaskRun to Pod and Task to TaskRun to
Pod.

This commit also adds a label whose key is pipeline.knative.dev/task
to all TaskRuns that refer to a Task with a TaskRef (the label is not
added to TaskRuns using an embedded TaskSpec) that contains the name of
the Task.

Fixes #501

## Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/knative/build-pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/knative/build-pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/knative/build-pipeline/blob/master/CONTRIBUTING.md#commit-messages)

## Release Notes

```release-note
* Labels on Tasks are now propagated to TaskRuns referring to that Task.
* TaskRuns referring to a Task will receive an additional label named pipeline.knative.dev/task containing the name of the Task.
* Labels on Pipelines are now propagated to PipelineRuns referring to that Pipeline.
```

### CODE REVIEW

File: docs/pipelines.md [hunk 1]
```diff
@@ -201,6 +202,26 @@ The resource `my-image` is expected to be given to the `deploy-app` `Task` from
 the `build-app` `Task`. This means that the `PipelineResource` `my-image` must
 also be declared as an output of `build-app`.
 
+## Labels
```
Comment (dwnusbaum) [inline docs/pipelines.md]: @bobcatfish At this point all of `Task`, `TaskRun`, `Pipeline`, and `PipelineRun` have a `Labels` section with a lot of duplicated info. Maybe better to create a new file to avoid the repetition? Thoughts?

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -171,15 +171,16 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
-	if equality.Semantic.DeepEqual(original.Status, pr.Status) {
-		// If we didn't change anything then don't call updateStatus.
+	if equality.Semantic.DeepEqual(original.Status, pr.Status) &&
+		reflect.DeepEqual(original.ObjectMeta.Labels, pr.ObjectMeta.Labels) {
+		// If we didn't change anything then don't call updateStatusAndLabels.
 		// This is important because the copy we loaded from the informer's
 		// cache may be stale and we don't want to overwrite a prior update
 		// to status with this stale state.
-	} else if _, err := c.updateStatus(pr); err != nil {
+	} else if _, err := c.updateStatusAndLabels(pr); err != nil {
```
Comment (dwnusbaum) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I'm not sure about this and the equivalent in `taskrun.go`. Before #367, both of these places were using `.Update` in their corresponding `updateStatus` functions, and I don't know if switching from `.UpdateStatus` back to `.Update` will break anything that that PR was doing. Would it be better to update the PipelineRun's labels immediately in the reconciler? Any ideas of how to avoid recomputing labels? For consistency it would be nice to actually set the labels on the `PipelineRun`/`TaskRun` but it feels a little awkward to update parts of those objects that are not the status.

File: test/builder/pod.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+// PodOp is an operation which modifies a Pod struct.
+type PodOp func(*corev1.Pod)
+
+// PodSpecOp is an operation which modifies a PodSpec struct.
+type PodSpecOp func(*corev1.PodSpec)
+
+// Pod creates a Pod with default values.
+// Any number of Pod modifiers can be passed to transform it.
+func Pod(name, namespace string, ops ...PodOp) *corev1.Pod {
```
Comment (dwnusbaum) [inline test/builder/pod.go]: I've added a builder for `Pods` since it was getting pretty hard to read what was going on in `taskrun_test.go`. Happy to back this out or split it into a separate PR if desired.

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -264,903 +277,307 @@ func TestReconcile(t *testing.T) {
...
-				Name:      "home",
-				ReadOnly:  false,
-				MountPath: "/builder/home",
-			},
-			},
-		}},
+		name:    "success",
+		taskRun: taskRunSuccess,
+		wantPod: tb.Pod("test-taskrun-run-success-pod-123456", "foo",
+			tb.PodAnnotation("sidecar.istio.io/inject", "false"),
```
Comment (dwnusbaum) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: Do we actually care about this Istio-related annotation, or is it just a holdover from when we were translating to Build and build was translating to Pod? If we aren't using it then maybe we should go ahead and delete it now.

**Comment by dwnusbaum** [comment]
Looks like the test failure is legitimate, I'm working on fixing it.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (dwnusbaum) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I think this is the change causing the tests to fail. Using `Update` does not update the `Status` even if the status field has changed, you have to use `UpdateStatus` to change the status. It looks like this is explained in [the docs for CRDs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource): 

> [If the status subresource is enabled] PUT/POST/PATCH requests to the custom resource ignore changes to the status stanza

I'll update the code to use `Update` to modify labels and leave `Status` handling as it was before. 

Also I noticed that #516 switched from `UpdateStatus` to `Update` in `cancel.go`, which might run into the same problem (or maybe it was a fix because the status in that code is not the actual status subresource since it is on the TaskRun's spec rather than TaskRun itself?).

Comment (pivotal-nader-ziada) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: #516 is only changing this for the cancel case which is different because the status in that code is not the actual status subresource like you said

File: docs/pipelines.md (hunk 1 above)
Comment (dwnusbaum) [reply docs/pipelines.md]: Done, moved to `labels.md`.

File: test/builder/pod.go (hunk 3 above)
Comment (dwnusbaum) [reply test/builder/pod.go]: @pivotal-nader-ziada No problem, I will take a look at splitting it out into another PR tomorrow!

**Comment by dwnusbaum** [comment]
Ok, I've removed the changes related to the Pod builder for the tests and will create a separate PR for that.

File: test/builder/pod.go (hunk 3 above)
Comment (pivotal-nader-ziada) [reply test/builder/pod.go]: thanks!

**Comment by pivotal-nader-ziada** [comment]
Thanks @dwnusbaum for making the changes!

/lgtm
/approve
